replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/WorkerManager.java

```diff
@@ -0,0 +1,148 @@
+   *
+   * @param broker the queue broker for retrieving queues
+   * @param siteManager the site data access manager
+   * @param nodeAdapters node adapters factory
+   * @param properties replication system properties
+   */
+  public WorkerManager(
+      QueueBroker broker,
+      SiteManager siteManager,
+      NodeAdapters nodeAdapters,
+      ReplicationProperties properties) {
+    this(
+        siteManager,
+        properties,
+        new WorkerThreadPoolFactory(
+            broker, siteManager.get(properties.getLocalSite()), siteManager, nodeAdapters));
+  }
+
+  /**
+   * Creates a new worker manager.
+   *
+   * @param siteManager the site data access manager
+   * @param properties replication system properties
+   */
+  public WorkerManager(
+      SiteManager siteManager,
+      ReplicationProperties properties,
+      WorkerThreadPoolFactory threadPoolFactory) {
+    this.siteManager = siteManager;
+    this.siteIds = Set.of(properties.getSites().toArray(new String[0]));
+    this.localSite = siteManager.get(properties.getLocalSite());
+    this.threadPoolFactory = threadPoolFactory;
+  }
+
+  /** Initializes this manager and begins monitoring of sites. */
+  public void init() {
+    executorService.scheduleAtFixedRate(this::monitorSites, 0L, 30L, TimeUnit.SECONDS);
+  }
+
+  /** Shutdown the manager and any workers that it is currently managing */
+  public void destroy() {
+    executorService.shutdownNow();
+    threadPools.forEach((site, pool) -> pool.shutdown());
+    threadPools.clear();
+  }
+
+  @VisibleForTesting
+  void monitorSites() {
+    Stream<Site> supportedSites =
+        siteManager
+            .objects()
+            .filter(site -> !SiteType.UNKNOWN.equals(site.getType()))
+            .filter(site -> !SiteKind.UNKNOWN.equals(site.getKind()))
+            .filter(site -> !site.getId().equals(localSite.getId()));
+
+    if (siteIds.isEmpty()) {
+      supportedSites.forEach(this::monitorSite);
+    } else {
+      supportedSites.filter(site -> siteIds.contains(site.getId())).forEach(this::monitorSite);
+    }
+
+    // if a site has been removed from storage, cleanup its thread pool
+    Set<String> threadPoolSites = new HashSet<>(threadPools.keySet());
+    siteManager.objects().forEach(site -> threadPoolSites.remove(site.getId()));
+    threadPoolSites.forEach(this::shutdownThreadPool);
+  }
+
+  private void monitorSite(Site site) {
+    final String siteId = site.getId();
+
+    final int poolSize =
+        Math.min(localSite.getParallelismFactor().orElse(1), site.getParallelismFactor().orElse(1));
+    WorkerThreadPool threadPool = threadPools.get(siteId);
+    if (threadPool != null) {
+      threadPool.setSize(poolSize);
+    } else {
+      threadPool = threadPoolFactory.create(siteId, poolSize);
+      threadPools.put(siteId, threadPool);
+    }
+  }
+
+  private void shutdownThreadPool(String siteId) {
+    WorkerThreadPool threadPool = threadPools.get(siteId);
```

Comment:
```suggestion
    WorkerThreadPool threadPool = threadPools.remove(siteId);
```

Then you don't need to remove it after. You end up accessing the collection once instead of twice.